lib/simplecov/branches_per_file.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module SimpleCov
+  #
+  # Get coverage report on certain file
+  #
+  module BranchesPerFile
+    #
+    # @param [String] source_file_path
+    #
+    # @return [Hash]
+    #
+    def self.start(source_file_path)
+      return {} unless SimpleCov.branchable_report
+      Coverage.start(:all)
+      require source_file_path
```

Comment:
Are you sure that the `require` is stable? I am receiving a loooot of `LoadError` errors when this is executed

Reply:
Well, thank you for being interested in this PR :), l had a look on your comments: 
1 - In line 16 `require` is called on the absolute file path so it's safe for the file to be required. `require` can through and LoadError if file/module target is not located in global load path constant.  
2 - About the non-nested style of defining modules, it should through another type of errors `NameError`.
3 - The logs that provided me with is `branches_per_file.rb:17:in `start': undefined method `[]'` the line `17`. This type of error is a result of totally empty ruby file which it's coverage report return an empty hash `{}`, **I will fix it**.
4 - Are you using `track_file` configuration on simplecov? could you please try to require the file that you have and with absolute path of it in the irb console? (This will help me to better understand the issue)
5-  `BranchesPerFile` concern the files that are missing any coverage report (not covered by applied tests), so it will return just the branches with zero coverage result. 
As a part of the solution on this case, is to write something similar to `LinesClassifier` that will parse the lines and give the report of branches, but this will make this PR very complicated to review. Better to have a simple solution that will return empty hash for now, and make a new PR with the new functionality of text parsing after this one will be accepted.
I appreciate your interest in this PR, hope our collaboration will make SimpleCov have this feature soon.

Reply:
1: 
> require is called on the absolute file path so it's safe

I do not quite understand what the file path has to do with safety of the file. My point was that with this change there is an unexpected eager loading on the execution of the test suite

3: This error appears also in a different case: execution of only a specified test or a specified directory

Example:
```
rspec spec/models

Randomized with seed 24924
..................................................

Finished in 1.89 seconds (files took 2.47 seconds to load)
50 examples, 0 failures

Randomized with seed 24924

Traceback (most recent call last):
	11: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/defaults.rb:27:in `block in <top (required)>'
	10: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:182:in `run_exit_tasks!'
	 9: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/configuration.rb:191:in `block in at_exit'
	 8: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:70:in `result'
	 7: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:310:in `process_coverage_result'
	 6: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:299:in `result_with_not_loaded_files'
	 5: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `add_not_loaded_files'
	 4: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `each'
	 3: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:265:in `block in add_not_loaded_files'
	 2: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:9:in `start'
	 1: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:26:in `call'
/Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/branches_per_file.rb:17:in `start': undefined method `[]' for nil:NilClass (NoMethodError)

```

4: No, am not using `track_file`

Sorry, I don't have time for this at the day-to-day work anymore; so I cannot get more into the details what is wrong there.

However, I made some of the problematic examples that came to my mind in my (much simple) pet project.

Have a look at https://github.com/klyonrad/lazybar/commits/simplecov-branch-cov

the full `rspec` run actually works fine with the edge cases that I wrote but the the execution of a directory or a single file breaks - with a different error then what I copied in my previous example:

```
rspec spec/models/cocktail_recipe_spec.rb

Randomized with seed 20614
..................

Top 2 slowest examples (1.15 seconds, 55.6% of total time):
  CocktailRecipe validations is not valid with only one ingredient
    1.05 seconds ./spec/models/cocktail_recipe_spec.rb:21
  CocktailRecipe methods about liking can show the users that like it
    0.09874 seconds ./spec/models/cocktail_recipe_spec.rb:70

Finished in 2.07 seconds (files took 3.19 seconds to load)
18 examples, 0 failures

Randomized with seed 20614

Traceback (most recent call last):
	13: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/defaults.rb:27:in `block in <top (required)>'
	12: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:182:in `run_exit_tasks!'
	11: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/configuration.rb:191:in `block in at_exit'
	10: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:70:in `result'
	 9: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:310:in `process_coverage_result'
	 8: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:299:in `result_with_not_loaded_files'
	 7: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `add_not_loaded_files'
	 6: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:263:in `each'
	 5: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov.rb:265:in `block in add_not_loaded_files'
	 4: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:9:in `start'
	 3: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/run_file_coverage.rb:26:in `call'
	 2: from /Users/luka/.rvm/gems/ruby-2.5.3/bundler/gems/simplecov-52827cb42b93/lib/simplecov/branches_per_file.rb:16:in `start'
	 1: from /Users/luka/development/lazy_bar/app/models/never_tested.rb:1:in `<top (required)>'
/Users/luka/development/lazy_bar/app/models/never_tested.rb:2:in `<class:NeverTested>': uninitialized constant NeverTested::NotExistingNamespace (NameError)
```

As I said: even though it would be a good thing to catch all these type of stupid programmer errors; I think an autoloading of all code is a bit unexpected behaviour

Reply:
@PragTob  see my comment above if you still not agree I will remove this class, and add it to my TODO list for next PR :)

Reply:
I suggest remove this class, because we tested your PR and found that took 10-15 seconds for our big legacy project... After remove this class everything working fast.

Reply:
@stepozer  Yep, that's what exactly I will do. We already agreed to do it at the comments above.